[clang] Fix nondeterminism in MemberPointerType

emaxx-google opened the pull request:
This commit fixes the nondeterminism issue in C++ header module enabled builds which were observed after
https://github.com/llvm/llvm-project/pull/132401.

The issue was related to the fact that the hash set operation in MemberPointerType::Profile() was triggering getMostRecentDecl(). As the latter may trigger the loading of new entities from the external AST source, this
was presumably causing reentrant modification of data structure or some other issue that affects
compiler's output in a nondeterministic way (likely depending on specific values hashes/pointers have).

The change should otherwise be a no-op, because whether we take a "most recent" or "any" Decl shouldn't
matter since `getCanonicalDecl()` is called on it anyway inside `MemberPointerType::Profile()`.

We haven't been able to come up with a deterministic regression test for this fix.

emaxx-google committed:
[clang] Fix nondeterminism in MemberPointerType

This commit fixes the nondeterminism issue in C++ header module enabled
builds which were observed after
https://github.com/llvm/llvm-project/pull/132401.

The issue was related to the fact that the hash set operation in
MemberPointerType::Profile() was triggering getMostRecentDecl(). The
root cause seems to be that the latter was leading to the reentrant
modification of the hash set, with some probability (likely depending
on the actual values of hashes).

We haven't been able to come up with a deterministic regression test for
this fix.

shafik reviewed (commented):
It is not clear to me why this is a specifically a problem when called from `Profile(...)` can you elaborate a bit more in the summary.

Does `getMostRecentCXXRecordDecl(...)` require a comment that documents this behavior so that future users do not fall into the same issue?

erichkeane commented on clang/include/clang/AST/Type.h:
```diff
@@ -3610,7 +3611,7 @@ class MemberPointerType : public Type, public llvm::FoldingSetNode {
 
   void Profile(llvm::FoldingSetNodeID &ID) {
-    Profile(ID, getPointeeType(), getQualifier(), getMostRecentCXXRecordDecl());
+    Profile(ID, getPointeeType(), getQualifier(), getCXXRecordDecl());
```
Is the problem that we run this profile in the 'middle' of a TU and the most recent decl changes by the end?  If so, this makes sense.

That said, I'd expect this to be the above suggestion in case this member pointer type has a redeclaration that perhaps changes what the actual referenced decl is.

```suggestion
    Profile(ID, getPointeeType(), getQualifier(), getCXXRecordDecl()->getCanonicalDecl());
```

emaxx-google replied:
Yes, but this suggestion seems not needed because this `getCanonicalDecl()` call is already made inside the `Profile()` method: https://github.com/llvm/llvm-project/blob/e8ae77947154e10dbc05cbb95ec9e10d3b0be13e/clang/lib/AST/Type.cpp#L5279

erichkeane replied:
Hmm... then I'm surprised that this patch is needed at all for the purposes of non-determinism.  `getMostRecentCXXRecordDecl()->getCanonicalDecl()` will get the same declaration as `getCXXRecordDecl()->getCanonicalDecl()`.

emaxx-google replied:
The issue is transient - the reproducibility rate on the reproducer (https://pastebin.com/6aL6rmBe) is only 1%. Also it's not the pointer we have here that's problematic, but rather the fact of extra AST node loading inside the `FoldingSet` hash calculation operation.

ilya-biryukov replied:
> Hmm... then I'm surprised that this patch is needed at all for the purposes of non-determinism

My theory is that non-determinism is caused by side-effects in the deserialization logic that runs behind `getMostRecentCXXRecordDecl()`. We probably just get unlucky with rehashing points at different times for the hash table (`FoldingSet`) that makes us either call or not call `getMostRecentCXXRecordDecl`.

It is not a new issue, just rare and now getting surfaced by the new place where `getMostRecentCXXRecordDecl` is called.

emaxx-google committed:
add comment

emaxx-google commented:
> It is not clear to me why this is a specifically a problem when called from `Profile(...)` can you elaborate a bit more in the summary.

I've added a bit more details - it's basically the fact that the "get most recent" operation triggered the load of new nodes from the PCM file. If all this happens while a hash for another node is calculated in `Profile()`, it means we're modifying the hash map while doing another operation with it.

At least that's the current theory, based on debugging the "deviating" executions of the compiler and seeing the `getMostRecentNonInjectedDecl()` call and the deserialization all during the `FoldingSet::NodeEquals()` function call.

> Does `getMostRecentCXXRecordDecl(...)` require a comment that documents this behavior so that future users do not fall into the same issue?

Sure, I've added a comment.

erichkeane reviewed (approved):
I'm still not sure I understand the problem this is solving, but I also don't see any harm in it, so I'm leaning towards accept here.  Aaron can be the final approval though.

mizvekov reviewed (commented):
I won't have time to review this closely until next week.

I think the 'inside profile' is a red herring here, this looks more like a problem from calling it from within deserialization.

We could try getting to the bottom of it and see if we can assert in the problematic cases, to avoid future repeats of the problem.

But the fix here is fine on its own, we don't need the most recent decl in most cases, and we kept one getter only for simplicity.

Though I think the name of the new getter needs more thought, it needs to avoid confusion with similar methods from the base type class.

Though I won't have time to give input on that until the weekend at least.

ilya-biryukov commented on clang/include/clang/AST/Type.h:
```diff
@@ -3602,6 +3602,10 @@ class MemberPointerType : public Type, public llvm::FoldingSetNode {
   }
 
   NestedNameSpecifier *getQualifier() const { return Qualifier; }
+  CXXRecordDecl *getCXXRecordDecl() const;
```
Could you make this private for now to give some space for @mizvekov to have a discussion for the proper name of the getter, while still being able to land this?

emaxx-google replied:
Sure, done. @mizvekov I'll also create a follow-up PR to rename the method - we can discuss the most appropriate choice there.

emaxx-google committed:
make new accessor private

emaxx-google commented:
> I'm still not sure I understand the problem this is solving, but I also don't see any harm in it, so I'm leaning towards accept here. Aaron can be the final approval though.

@AaronBallman I'd like to proceed with landing this as a short-term fix for the nondeterminism issue we're hitting at Google, but if you have any additional feedback I'm happy to address it in a follow-up. (We're also very interested in investigating the actual underlying issues that lead to these subtle problems, but it'll probably take much more time.)

AaronBallman reviewed (approved):
I think this is fine as a temporary hack, but please add a FIXME comment explaining that this is working around a deeper underlying issue.

emaxx-google committed:
add FIXME

ilya-biryukov reviewed (approved):
Approving to land this and unblock our internal releases.
Obviously, I encourage to address the issues raised by other reviewers in the follow-ups.

In particular, the most easily actionable one is about the naming of the getter from @mizvekov . Let's wait for the comments and fix that quickly if a better name is suggested.

emaxx-google commented:
Thanks everyone!
